src/index/src/inverted_index/format/writer.rs

```diff
@@ -23,19 +25,27 @@ use crate::inverted_index::error::Result;
 pub use crate::inverted_index::format::writer::blob::InvertedIndexBlobWriter;
 use crate::inverted_index::Bytes;
 
+pub type ValueStream = Box<dyn Stream<Item = Result<(Bytes, BitVec)>> + Send + Unpin>;
+
 /// Trait for writing inverted index data to underlying storage.
+#[mockall::automock]
 #[async_trait]
-pub trait InvertedIndexWriter {
+pub trait InvertedIndexWriter: Send {
     /// Adds entries to an index.
     ///
     /// * `name` is the index identifier.
     /// * `null_bitmap` marks positions of null entries.
     /// * `values` is a stream of values and their locations, yielded lexicographically.
     ///    Errors occur if the values are out of order.
-    async fn add_index<S>(&mut self, name: String, null_bitmap: BitVec, values: S) -> Result<()>
-    where
-        S: Stream<Item = Result<(Bytes, BitVec)>> + Send + Unpin;
+    async fn add_index(
+        &mut self,
+        name: String,
+        null_bitmap: BitVec,
+        values: ValueStream,
+    ) -> Result<()>;
 
     /// Finalizes the index writing process, ensuring all data is written.
-    async fn finish(&mut self) -> Result<()>;
+    /// `total_row_count` and `segment_row_count` is used to fill in the metadata.
+    async fn finish(&mut self, total_row_count: u64, segment_row_count: NonZeroUsize)
```

Comment:
I have a little question about those new in-params. Tracking row counts in writer seems more intuitive to me. Why add them explicitly?

Reply:
> Tracking row counts in writer

Writer cannot the track row count. It can only track the number of distinct values.